### ISSUE TITLE
Wrong `minion_id` in matcher methods.

### ISSUE THREAD

**Issue by funbaker** [saltstack/salt#55656]
### Description of Issue

When filtering pillars with `match.filter_by`, they don't get matched correctly.
Reason is a wrong `minion_id` in the matcher methods.
This bug was introduced in https://github.com/saltstack/salt/pull/48809/ , hence the question if that was somewhat intentional.

Can provide PR after clarification.

### Setup
Any Setup with minion-filtered pillars, as described in https://docs.saltstack.com/en/latest/ref/modules/all/salt.modules.match.html

### Steps to Reproduce Issue
Setup minion-filtered pillars as described in https://docs.saltstack.com/en/latest/ref/modules/all/salt.modules.match.html

### Versions Report
Most of the 2019 versions, introduced with the PR https://github.com/saltstack/salt/pull/48809/

**Comment by Ch3LL** [comment]
just a carification. are you refering to this setup in the docs? Its not clear to me exactly your setup when running into this issue:

```
Pillar Example:

# Filter the data for the current minion into a variable:
{% set roles = salt['match.filter_by']({
    'web*': ['app', 'caching'],
    'db*': ['db'],
}, default='web*') %}

# Make the filtered data available to Pillar:
roles: {{ roles | yaml() }}
```

???

**Comment by funbaker** [comment]
@Ch3LL yes that's it, with the little difference that i use no wildcard yet.

**Comment by Ch3LL** [comment]
I'm having a hard time replicating any buggy behavior. Can you share a complete use case where your seeing the wrong minion_id appear?

**Comment by funbaker** [comment]
> I'm having a hard time replicating any buggy behavior. Can you share a complete use case where your seeing the wrong minion_id appear?

```
{% set admins = salt.match.filter_by({
    'host1': ['localadmin1'],
    '*': ['admin1', 'admin2']
}) %}

```

In 2018.3 it works as expected with `admin1`, `admin2` and `localadmin1`.
In 2019.2 it doesn't match at all.

It broke in the PR linked above, and i already know how to fix it.
The responsible line is https://github.com/saltstack/salt/pull/48809/files#diff-889b4ad7294d8f45dfd1c3a87b9ba427L45

I'm not sure though why this line was removed, so I need feedback on that.

**Comment by Ch3LL** [comment]
your matcher is still working for me. Anyways looking at that line it looks like it was removed to grab `id` from `__opts__` instead of `__grains__` if `minion_id` is not set, which does work for me.

**Comment by Ch3LL** [comment]
maybe the author @cro can chime in here as well to explain the change.

**Comment by Ch3LL** [comment]
bump @cro

**Comment by peter-slovak** [comment]
We're running into the same problem. Although getting `minion_id` from `__opts__` should be a safe fallback from `grains['id']`, the reality is that `__opts__['minion_id']` actually contains a different minion's name. 

That's why the `pcre_match` keeps working and `glob_match` broke down - the first one uses `opts['id']` (which in our case contains the correct minion name), the other one uses `__opts__['minion_id']` (which contains another minion's name).

I'm not sure how this is possible; my guess would be that `__opts__['minion_id']` is somehow, somewhere modified as a result of a race or thread unsafety. 

This problem goes away for a while after we restart the master. However, with 100% reproducibility, after we render the pillar on all minions (`salt --async '*' pillar.items`), the problem is back.

**Comment by peter-slovak** [comment]
FWIW, changing `minion_id = opts.get('minion_id', opts['id'])` to `minion_id = opts['id']` in `matchers/glob_match.py` works around the problem (bypassing the incorrect value in `opts['minion_id']`.

**Comment by dwoz** [comment]
Fix merged and shipped in a supported release. Closing per bulk triage. Reopen if this reproduces on a currently supported release (3006.x, 3007.x, 3008.x, master).
